### DISCUSSION TITLE
Pre-render: julia.jl scripts

### DISCUSSION THREAD

**Discussion by Kangaa** [quarto-dev/quarto-cli#4451] [Q&A] [answered]
Just wondering whether I'm missing something or whether it is actually not possible to run a Julia script using the 'pre-render' field in a  _quarto.yml? 
If this isn't possible, it would be a much welcome feature!

**Comment by cscheid** [comment, marked as answer]
You're correct, it's currently not possible to do so. The support for pre-render scripts goes through the same code path as `quarto run`, and `quarto run` does not support Julia.

`quarto run` only supports the languages necessary for quarto's engines and its dependencies: `knitr` requires R, `jupyter` requires Python, and the base "non-engine" `markdown` requires deno (so we support typescript). Finally, we support Lua since Pandoc does so.

Julia is very likely to be our third most popular code language, but we are reluctant to add general `quarto run` support to languages not required by our engines, mostly because it's not clear where we should stop. 

I agree that there's a strong case for Julia based on its popularity! But note that we support `haskell`,  `go`, and `apl` code cells through jupyter kernels. I don't think we should support those as pre-render scripts, and Julia falls in that category as well.

I think we could change our position, but we currently have no plans to add this feature.

**Reply by baptiste** [reply]
I may have misunderstood something, but the docs'

> Note that pre-render and post-render also support arbitrary shell commands.

suggests to me that `pre-render: julia script.jl` would work, provided `julia` is registered in the user's path?

**Reply by cscheid** [reply]
Yes, that would work :) I should have clarified, and I was giving the answer for `pre-render: script.jl`.

For us, the distinction is important. In the other languages, we attempt to match things like the environment (we try to make sure that the version of R used for the script is the same version of R used for execution).

In the case of `julia script.jl`, there's (hopefully clearly) no promise that we'll know how to resolve the right version of julia, etc. And that's also equivalent to wrapping the call in a python script.

**Comment by jgunstone** [comment]
is it possible (/could it be) to run shell scripts (`script.sh`) as a pre-render script?

**Reply by mcanouil** [reply]
See <https://quarto.org/docs/projects/scripts.html#pre-and-post-render>.

> Note that pre-render and post-render also support arbitrary shell commands. So you could for example use make to do data preparation this way:

**Reply by jgunstone** [reply]
great thanks - yeah I thought I read this previously and then got confused as only ts, py, lua and R were explicitly described. 
thanks for the hint - works beautifully

**Reply by jgunstone** [reply]
sorry to revisit this - 
it was working as described in the docs, then after I restarted my machine I've had a strange problem with this. 

below I've copied my console that demonstrates: 
1. running the shell script successfully from the command line
2. the contents of my `_quarto.yml` (with the shell script set as a pre-render script)
3. the traceback when I try to run `quarto preview .` where it says that the shell script I just ran doesn't exist... 

the shell script is defo in the same folder as the `_quarto.yml`... so I'm not sure why to try... any ideas? 

many thanks


```bash
(meetings-and-notes) (base) jovyan@PORT4028:~/project-tracker$ ./get_gh_project_data.sh 
retrieved project12 data
(meetings-and-notes) (base) jovyan@PORT4028:~/project-tracker$ cat _quarto.yml
project:
  type: website
  pre-render: 
    - get_gh_project_data.sh
    - process_gh_project_data.py
    - get_webapp_data.py
    - process_webapp_data.py
    - build_project_pages.py

website:
  title: "Project-tracker"
  page-navigation: true
  sidebar:
    style: "docked"
    search: true
    contents:
      - section: "Projects"
        href: index.qmd
        contents:
          - auto: "projects/*.qmd"

format:
  html:
    theme: cosmo
    css: styles.css
    toc: true
    grid:
(meetings-and-notes) (base) jovyan@PORT4028:~/project-tracker$ quarto preview .
Preparing to preview
get_gh_project_data.sh
ERROR: Error executing 'get_gh_project_data.sh': No such file or directory (os error 2)

Stack trace:
    at execProcess (file:///home/jovyan/miniforge3/envs/meetings-and-notes/bin/quarto.js:5297:15)
    at runScripts (file:///home/jovyan/miniforge3/envs/meetings-and-notes/bin/quarto.js:77655:35)
    at runPreRender (file:///home/jovyan/miniforge3/envs/meetings-and-notes/bin/quarto.js:77632:11)
    at renderProject (file:///home/jovyan/miniforge3/envs/meetings-and-notes/bin/quarto.js:77356:15)
    at eventLoopTick (ext:core/01_core.js:183:11)
    at async serveProject (file:///home/jovyan/miniforge3/envs/meetings-and-notes/bin/quarto.js:97496:24)
    at async Command.fn (file:///home/jovyan/miniforge3/envs/meetings-and-notes/bin/quarto.js:98175:9)
    at async Command.execute (file:///home/jovyan/miniforge3/envs/meetings-and-notes/bin/quarto.js:8104:13)
    at async quarto (file:///home/jovyan/miniforge3/envs/meetings-and-notes/bin/quarto.js:114999:5)
    at async file:///home/jovyan/miniforge3/envs/meetings-and-notes/bin/quarto.js:115017:9
(meetings-and-notes) (base) jovyan@PORT4028:~/project-tracker$ 
```

**Reply by mcanouil** [reply]
If you don't provide us with the means to reproduce your issue, there is unfortunately little or even nothing we can do to help.

**Reply by jgunstone** [reply]
hi - yes I understand... 
I think the issue is with my installation either way... 
I just started making a quickstart example to explain in better, but within that example it worked! 

I then realised the following: 
- If I navigate to the a folder on the cmd line and open vs code from that dir (i.e. `code .`)
  - then within the terminal runL `quarto preview .`  -> it works
- If I just navigate to the a folder on the cmd line and then just run the command `quarto preview .`
  - it doesn't work

either way seems like my issues rather than yours and I have a method now... 
thanks
